File: huffman_compression.py
codes = {}

def frequency(str):
    freqs = {}
    for ch in str:
        freqs[ch] = freqs.get(ch,0) + 1
    return freqs

freqs = frequency("javascript")
print freqs

def sortFreq(freqs):
    letters = freqs.keys()
    tuples = []
    for let in letters:
        tuples.append((freqs[let],let))
    tuples.sort()
    return tuples

sort = sortFreq(freqs)
print sort

def buildTree(tuples):
    while len(tuples) > 1:
        leastTwo = tuple(tuples[0:2])
        theRest = tuples[2:]
        combFreq = leastTwo[0][0] + leastTwo[1][0]
        tuples = theRest + [(combFreq,leastTwo)]
        tuples.sort()
    return tuples[0]

build = buildTree(sort)
print build

def trimTree(tree):
    p = tree[1]
    if type(p) == type(""): return p
    else: return (trimTree(p[0]), trimTree(p[1]))

trim = trimTree(build)
print trim

def assignCodes(node, pat=""):
    global codes
    if type(node) == type(""):
        codes[node] = pat
    else:
        assignCodes(node[0], pat+"0")
        assignCodes(node[1], pat+"1")

assignCodes(trim)
print codes

def encode(str):
    global codes
    output = ""
    for ch in str:
        output += codes[ch]
    return output

encoding = encode("javascript")
print encoding

def decode(tree,str):
    output = ""
    p = tree
    for bit in str:
        if bit == '0': p = p[0]
        else         : p = p[1]
        if type(p) == type(""):
            output += p
            p = tree
    return output

decoding = decode(trim, encoding)
print decoding

File: huffman_compression.js
var codes = {}

function frequency(str)
{
    var freqs = {};
    for (var count = 0; count < str.length; count++)
    {
        if (str[count] in freqs)
            freqs[str[count]] += 1;
        else
            freqs[str[count]] = 1;
    }
    return freqs;
}

var freqs = frequency("javascript");
console.log(freqs);

function sortFreq(freqs)
{
    var tuples = []
    for (var letter in freqs)
        tuples.push([freqs[letter],letter]);
    return tuples.sort();
}

var tuples = sortFreq(freqs)
console.log(tuples);

function buildTree(tuples)
{
    while (tuples.length > 1)
    {
        var leastTwo = [tuples[0][1], tuples[1][1]];
        var theRest = tuples.slice(2);
        var combFreq = tuples[0][0] + tuples[1][0];
        var tuples = theRest;
        var end = [combFreq, leastTwo];
        tuples.push(end);
        tuples.sort();
    }
    return tuples[0][1];
}

var tree = buildTree(tuples);
console.log(tree);

var pat = "";
function assignCodes(node, pat)
{
    if (typeof(node) == typeof(""))
        codes[node] = pat;
    else
    {
        assignCodes(node[0], pat+"0");
        assignCodes(node[1], pat+"1");
    }
}

assignCodes(tree, pat);
console.log(codes);

function encode(str)
{
    var output = "";
    for (var ch in str)
        output += codes[str[ch]];
    return output;
}

var encoding = encode("javascript");
console.log(encoding);

function decode(tree,enStr)
{
    var output = "";
    var p = tree;
    for (var bit in enStr)
    {
        if (enStr[bit] == "0")
            p = p[0];
        else
            p = p[1];
        if (typeof(p) == typeof(""))
        {
            output += p;
            p = tree;
        }
    }
    return output;
}

var decoding = decode(tree, encoding);
console.log(decoding);
